### PR TITLE
SNOW-857829 Fix username and password requiredness

### PR DESCRIPTION
**Pull request by sfc-gh-pfus**
### Description
Fixes https://github.com/snowflakedb/gosnowflake/issues/843
Snowflake required username and password in most auth types, even if they don't really needed. Fix enables username and password checking only if they are really required.

### Checklist
- [x] Code compiles correctly
- [x] Run ``make fmt`` to fix inconsistent formats
- [x] Run ``make lint`` to get lint errors and fix all of them
- [x] Created tests which fail without the change (if possible)
- [x] All tests passing
- [x] Extended the README / documentation, if necessary

### CODE REVIEW

File: dsn.go [hunk 1]
```diff
@@ -390,16 +390,11 @@ func fillMissingConfigParameters(cfg *Config) error {
 		return ErrEmptyAccount
 	}
 
-	if cfg.Authenticator != AuthTypeOAuth && strings.Trim(cfg.User, " ") == "" {
-		// oauth does not require a username
+	if authRequiresUser(cfg) && strings.Trim(cfg.User, " ") == "" {
```
Comment (sfc-gh-dheyman-1) [inline dsn.go:393]: `authRequiredUsername`?

Comment (sfc-gh-dheyman-1) [inline dsn.go:393]: i would suggest `strings.TrimSpace`

File: dsn.go [hunk 2]
```diff
@@ -390,16 +390,11 @@ func fillMissingConfigParameters(cfg *Config) error {
...
 		return ErrEmptyUsername
 	}
 
-	if cfg.Authenticator != AuthTypeExternalBrowser &&
-		cfg.Authenticator != AuthTypeOAuth &&
-		cfg.Authenticator != AuthTypeJwt &&
-		strings.Trim(cfg.Password, " ") == "" {
-		// no password parameter is required for EXTERNALBROWSER, OAUTH or JWT.
+	if authRequiresPassword(cfg) && strings.Trim(cfg.Password, " ") == "" {
```
Comment (sfc-gh-dheyman-1) [inline dsn.go:397]: `strings.TrimSpace`?

**Review by sfc-gh-dheyman-1 — APPROVED** [review]

File: dsn.go (hunk 1 above)
Comment (sfc-gh-pfus) [reply dsn.go:393]: I would stick to the "domain" name of this property, which is "user" in this project.

Comment (sfc-gh-pfus) [reply dsn.go]: Done.

File: dsn.go (hunk 2 above)
Comment (sfc-gh-pfus) [reply dsn.go]: Done.

File: dsn.go (hunk 1 above)
Comment (sfc-gh-dheyman-1) [reply dsn.go]: Alright. A bit misleading though.

File: dsn_test.go [hunk 3]
```diff
@@ -582,112 +582,172 @@ func TestParseDSN(t *testing.T) {
...
-			driverErrE, okE := test.err.(*SnowflakeError)
-			driverErrG, okG := err.(*SnowflakeError)
-			if okE && !okG || !okE && okG {
-				t.Fatalf("%d: Wrong error. expected: %v, got: %v", i, test.err, err)
-			}
-			if okE && okG {
-				if driverErrE.Number != driverErrG.Number {
-					t.Fatalf("%d: Wrong error number. expected: %v, got: %v", i, driverErrE.Number, driverErrG.Number)
+		t.Run("TestParseDSN", func(t *testing.T) {
+			// t.Logf("Parsing testcase %d, DSN: %s", i, test.dsn)
```
Comment (sfc-gh-pbulawa) [inline dsn_test.go:644]: I think we could remove the comment.

**Review by sfc-gh-pbulawa — APPROVED** [review]

File: dsn_test.go [hunk 4]
```diff
@@ -582,112 +582,171 @@ func TestParseDSN(t *testing.T) {
 		},
 	}
 
+	for _, at := range []AuthType{AuthTypeExternalBrowser, AuthTypeOAuth} {
+		testcases = append(testcases, tcParseDSN{
+			dsn: fmt.Sprintf("@host:777/db/schema?account=ac&protocol=http&authenticator=%v", strings.ToLower(at.String())),
+			config: &Config{
+				Account: "ac", User: "", Password: "",
+				Protocol: "http", Host: "host", Port: 123,
```
Comment (sfc-gh-dprzybysz) [inline dsn_test.go:590]: is it ok that port is different than in dsn? also the same in other tests?

Comment (sfc-gh-pfus) [reply dsn_test.go]: It doesn't matter, because earlier check fails. But to have order I fixed it. Thanks!

**Review by sfc-gh-dprzybysz — APPROVED** [review]

**Review by sfc-gh-igarish — APPROVED** [review]
LGTM

**Pull request merged**